File: cookies/sessionId.ts
import { ReadonlyRequestCookies } from "next/dist/server/web/spec-extension/adapters/request-cookies";
import { NextResponse } from "next/server";

const SESSION_ID_COOKIE = "toni-tunes-session-id";

export function setSessionIdCookie(res: NextResponse, sessionId: string) {
  res.cookies.set(SESSION_ID_COOKIE, sessionId, {
    httpOnly: true,
    secure: process.env.ENVIRONMENT !== "dev",
    maxAge: 60 * 60 * 24 * 7, // one week
    path: "/",
  });
}

export function getSessionIdCookie(
  cookieStore: ReadonlyRequestCookies,
): string | null {
  const cookie = cookieStore.get(SESSION_ID_COOKIE);

  if (!cookie) {
    console.log("cookie is null");
    return null;
  }

  console.log("cookie not null", cookie.value);
  return cookie.value;
}

File: cookies/session_id.go
package cookies

import (
	"net/http"
	"os"
	"time"
)

const COOKIE_LIFETIME = 7 * 24 * time.Hour
const SESSION_ID_COOKIE_NAME = "toni-tunes-session-id"

func SetSessionId(w http.ResponseWriter, sessionId string) {
	var secure bool
	if os.Getenv("ENVIRONMENT") == "dev" {
		secure = false
	} else {
		secure = true
	}
	cookie := http.Cookie{
		Name:     SESSION_ID_COOKIE_NAME,
		Value:    sessionId,
		Secure:   secure,
		HttpOnly: true,
		Path:     "/",
		Expires:  time.Now().Add(COOKIE_LIFETIME),
		SameSite: http.SameSiteStrictMode,
	}

	http.SetCookie(w, &cookie)
}

func GetSessionId(r *http.Request) string {
	cookie, err := r.Cookie(SESSION_ID_COOKIE_NAME)
	if err != nil {
		return ""
	}

	return cookie.Value
}

func RemoveSessionId(w http.ResponseWriter) {
	SetSessionId(w, "")
}
